### PR TITLE
MM-32950: Reliable WebSockets: Basic single server

### PR DESCRIPTION
**Pull request by agnivade**
This PR adds reliable websocket support for a single server.

Below is a brief overview of the three states of a connection:

Normal:
- All messages are routed via web hub.
- Each web conn has a send queue to which it gets pushed.
- A message gets pulled from the queue, and before it
gets written to the wire, it is added to the dead queue.

Disconnect:
- Hub Unregister gets called, where the connection is just
marked as inactive. And new messages keep getting pushed
to the send queue.

If it gets full, the channel is closed and the conn gets removed
from conn index.

Reconnect:
- We query the hub for the connection ID, and get back the
queues.
- We construct a WebConn reusing the old queues, or a fresh one
depending on whether the connection ID was found or not.
- Now there is a tricky bit here which needs to be carefully processed.
On register, we would always send the hello message in the send queue.
But we cannot do that now because the send queue might already have messages.

Therefore, we don't send the hello message from web hub, if we reuse a connection.

Instead, we move that logic to the web conn write pump. We check if
the sequence number is in dead queue, and if it is, then we drain
the dead queue, and start consuming from the active queue.
No hello message is sent here.

But if the message does not exist in the dead queue, and the sequence number
is actually something that should have existed, then we set
a new connction id and clear the dead queue, and send a hello message.
The client, on receiving a new connection id will automatically
set its sequence number to 0, and make the sync API calls to manage
any lost data.

https://mattermost.atlassian.net/browse/MM-32590

```release-note
NONE
```

### CODE REVIEW

**Comment by agnivade** [comment]
Webapp changes here: https://github.com/mattermost/mattermost-webapp/pull/7921

**Comment by agnivade** [comment]
I have tested this manually for all possible scenarios by manually tweaking the code to introduce errors.

Also tested in race mode, and ran a small load test to verify any edge cases.

**Review by streamer45 — COMMENTED** [review]
Looks very good :tada:. I really appreciated the thorough commenting.
Gave this a first pass and left a few comments.

File: app/web_conn.go [hunk 1]
```diff
@@ -304,19 +377,123 @@ func (wc *WebConn) writePump() {
...
+		if elem == nil {
+			return false
+		}
+
+		if elem.GetSequence() == wc.Sequence {
+			return true
+		}
+	}
+	return false
+}
```
[the remark this answers is not in the document]
Comment (agnivade) [reply app/web_conn.go:440]: Sure.

File: app/web_conn.go [hunk 2]
```diff
@@ -304,19 +377,123 @@ func (wc *WebConn) writePump() {
...
+			}
+		}
+		return nil
+	}
+
+	// We go on until next sequence number is smaller than previous one.
+	// Which means it has rolled over.
+	currPtr := wc.deadQueuePointer
+	for ; wc.deadQueue[currPtr].GetSequence() < wc.Sequence; currPtr = (currPtr + 1) % deadQueueSize {
+	}
```
[the remark this answers is not in the document]
Comment (agnivade) [reply app/web_conn.go:482]: Hmm .. good point. I'll check and update.

File: api4/websocket.go [hunk 3]
```diff
@@ -36,53 +37,75 @@ func connectWebSocket(c *Context, w http.ResponseWriter, r *http.Request) {
...
+				// or server restart. In that case, we set a new one.
+				connID = model.NewId()
+			} else {
+				// Connection is present, we get the active queue, dead queue
+				activeQueue = res.ActiveQueue
+				deadQueue = res.DeadQueue
+				deadQueuePointer = res.DeadQueuePointer
+				active = false
+				// Now we get the sequence number
+				seqVal := r.URL.Query().Get(sequenceNumberParam)
```
[the remark this answers is not in the document]
Comment (agnivade) [reply api4/websocket.go:74]: Oh yeah .. I thought long about this one. I settled down on this approach after my third iteration :P

The issue with reviving the queue in-place is that the registration code gets unnecessarily complicated. You will have to drain the dead queue _before_ starting to send the active queue. But hub can only send to the active queue. So the best compromise I came up was to initialize the webconn with either fresh data, or reused data, and clear out stale data if any. And then web hub can send as usual.

This way most complexity remains inside webconn, rather than spread all over the place.

I think this looks a bit odd because we are throwing the data all the way from hub to the api layer, and then coming back to app layer again. I didn't think too much about this because I have been looking at this code for way too long :P But I'll clear my mind and probably move some of the initialization step inside webconn. That way the overall logic should look a bit cleaner.

**Comment by agnivade** [comment]
> I really appreciated the thorough commenting.

Definitely, I needed the comments to keep track of things myself. If I didn't comment some thing, the next day, I would do the same thing, and then remember why I did not choose to do something.

I think the end result is more or less manageable. Initially, it looked like it would get way more complicated. :P

**Comment by agnivade** [comment]
@streamer45 - Take a look now. The API layer is significantly cleaned up.

File: app/web_conn.go [hunk 4]
```diff
@@ -136,6 +210,22 @@ func (wc *WebConn) SetConnectionID(id string) {
 	wc.connectionID.Store(id)
 }
 
+// GetConnectionID returns the connection id of the connection.
+func (wc *WebConn) GetConnectionID() string {
+	return wc.connectionID.Load().(string)
+}
+
+// areAllInactive returns whether all of the connections
+// are inactive or not
```
Comment (streamer45) [inline app/web_conn.go:219]:
```suggestion
// areAllInactive returns whether all of the connections
// are inactive or not.
```

File: app/web_conn.go [hunk 5]
```diff
@@ -249,8 +358,8 @@ func (wc *WebConn) writePump() {
 			buf.Reset()
 			var err error
 			if evtOk {
-				cpyEvt := evt.SetSequence(wc.Sequence)
-				err = cpyEvt.Encode(enc)
+				evt = evt.SetSequence(wc.Sequence)
+				err = evt.Encode(enc)
```
Comment (streamer45) [inline app/web_conn.go:362]:
```suggestion
				err = evt.SetSequence(wc.Sequence).Encode(enc)
```

File: app/web_conn.go [hunk 6]
```diff
@@ -304,19 +414,117 @@ func (wc *WebConn) writePump() {
...
 	wc.WebSocket.SetWriteDeadline(time.Now().Add(writeWaitTime))
 	return wc.WebSocket.WriteMessage(msgType, data)
 }
 
+func (wc *WebConn) writeMessage(msg *model.WebSocketEvent) error {
+	// We don't use the encoder from the write pump because it's unwieldy to pass encoders
+	// around, and this is only called during initialization of the webConn.
+	var buf bytes.Buffer
+	enc := json.NewEncoder(&buf)
+	err := msg.Encode(enc)
```
Comment (streamer45) [inline app/web_conn.go:429]:
```suggestion
	err := msg.Encode(json.NewEncoder(&buf))
```

File: app/web_conn.go [hunk 7]
```diff
@@ -304,19 +414,117 @@ func (wc *WebConn) writePump() {
...
+	if err != nil {
+		mlog.Warn("Error in encoding websocket message", mlog.Err(err))
+		return nil
+	}
+	wc.Sequence++
```
Comment (streamer45) [inline app/web_conn.go:434]: How does this relate to the increment in `writePump()`? Are they both necessary?

https://github.com/mattermost/mattermost-server/blob/54e1590b2bd276a8b3fcf19784556bf33ac92b12/app/web_conn.go#L363

File: app/web_conn.go [hunk 8]
```diff
@@ -304,19 +414,117 @@ func (wc *WebConn) writePump() {
...
+		if wc.deadQueue[deadQueueSize-1].GetSequence() == wc.Sequence-1 {
+			return false
+		}
+		return true
+	}
+	if wc.deadQueue[wc.deadQueuePointer-1].GetSequence() == wc.Sequence-1 {
+		return false
+	}
+	return true
+}
```
Comment (streamer45) [inline app/web_conn.go:461]: Interestingly I find this logic a bit hard to digest and I think it's mainly because we are doing two things at a time and returning in so many places. I am wondering if it would be clearer if we first find the element we are looking for (the last) and then do the sequence check (if not nil of course).

File: app/web_conn.go (hunk 7 above)
Comment (agnivade) [reply app/web_conn.go:434]: Yes absolutely. `wc.writeMessage` is only called before the for loop in the write pump actually starts.

Think of it this way:
- Client sends the last seen sequence number, which gets set in the constructor.
- We check for the dead queue, and either drain the queue, or send hello message. So sequence number needs to be incremented there.
- Then we come to our usual goroutine, where everything else remains same.

File: app/web_conn.go (hunk 8 above)
Comment (agnivade) [reply app/web_conn.go:461]: Ah ok, let me see what I can do.

File: app/web_conn.go (hunk 7 above)
Comment (streamer45) [reply app/web_conn.go:434]: Makes sense. I've got confused with the renaming and thought it was called further down in the function. All good then :+1:

File: app/web_conn.go [hunk 9]
```diff
@@ -304,19 +414,117 @@ func (wc *WebConn) writePump() {
...
+	return false, 0
+}
+
+func (wc *WebConn) clearDeadQueue() {
+	for i := 0; i < deadQueueSize; i++ {
+		if wc.deadQueue[i] == nil {
+			return
+		}
+		wc.deadQueue[i] = nil
+	}
```
Comment (isacikgoz) [inline app/web_conn.go:487]: If we are sure about `wc.deadQueuePointer`, could we use something like this?:
```suggestion
	copy(wc.deadQueue[:wc.deadQueuePointer], make([]*model.WebSocketEvent, wc.deadQueuePointer))
```

File: app/web_conn_test.go [hunk 10]
```diff
@@ -129,6 +129,170 @@ func TestWebConnDeadQueue(t *testing.T) {
...
+
+	run := func(seqNum int64, limit int) {
+		s := httptest.NewServer(handler(t, seqNum, limit))
+		defer s.Close()
+
+		wc := dialConn(t, th.App, s.Listener.Addr())
+		defer wc.WebSocket.Close()
+
+		var i int
+		for ; i < limit; i++ {
```
Comment (isacikgoz) [inline app/web_conn_test.go:272]: Is this the new style for initiating loop variables? :p

**Review by isacikgoz — APPROVED** [review]
Looks good to me, awesome work @agnivade 🎉 I have minor things but none of the are blocking.

File: app/web_conn_test.go (hunk 10 above)
Comment (agnivade) [reply app/web_conn_test.go:272]: lol no :joy: I was using `i` later, but then that code got removed. Will fix.

File: app/web_conn.go (hunk 9 above)
Comment (agnivade) [reply app/web_conn.go:487]: We can, but that creates extra allocations. Actually, we can even do `wc.deadQueue = wc.deadQueue[:0]`. :wink:

File: app/web_conn.go (hunk 5 above)
Comment (agnivade) [reply app/web_conn.go:362]: Actually, we shouldn't do this. `evt` is "immutable", and `SetSequence` returns a new event, which is used later in the code. So if we don't re-assign `evt`, then it will use the wrong sequence number.

File: app/web_conn.go [hunk 11]
```diff
@@ -53,27 +68,77 @@ type WebConn struct {
...
+	sessionToken atomic.Value
+	session      atomic.Value
+	connectionID atomic.Value
+	endWritePump chan struct{}
+	pumpFinished chan struct{}
+}
+
+// CheckConnResult indicates whether a connectionID was present in the hub or not.
+// And if so, contains the active and dead queue details.
+type CheckConnResult struct {
```
[the remark this answers is not in the document]
Comment (agnivade) [reply app/web_conn.go:87]: Ahh crap, it needs to be exported because the opentracing package is a sub-package of app, and therefore any struct returned by an app method _must_ be exported. :(

File: app/web_conn.go (hunk 5 above)
Comment (streamer45) [reply app/web_conn.go:364]: That's absolutely right :+1:

File: app/web_conn.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (streamer45) [reply app/web_conn.go:89]: Too bad :P

File: app/web_hub.go [hunk 12]
```diff
@@ -387,42 +419,91 @@ func (h *Hub) Start() {
...
+
+		connIndex := newHubConnectionIndex(inactiveConnReaperInterval)
 
 		for {
 			select {
 			case webSessionMessage := <-h.checkRegistered:
 				conns := connIndex.ForUser(webSessionMessage.userID)
 				var isRegistered bool
 				for _, conn := range conns {
+					if !conn.active {
```
Comment (streamer45) [inline app/web_hub.go:433]: Do all these operations (read/write) on `WebConn.active` need to be atomic?

**Review by streamer45 — COMMENTED** [review]
Awesome :+1: , just one question remaining :)

File: app/web_hub.go (hunk 12 above)
Comment (agnivade) [reply app/web_hub.go:433]: Nope. I have checked this with `-race` flag. The `active` bit is only set during initialization. And it's always read/written in the hub loop, which always runs as a single goroutine. Didn't want to unnecessarily bring more atomics.

Comment (streamer45) [reply app/web_hub.go:433]: Fair enough. Let's just be careful with this because now we have a mixed (getters/setters and direct) way of accessing connection fields based on where in the code we use them so it can get tricky and error prone.

**Review by streamer45 — APPROVED** [review]
Looks great. Brilliant work :+1:

File: app/web_hub.go (hunk 12 above)
Comment (agnivade) [reply app/web_hub.go:433]: Fair point. I am assuming this part of the codebase is so sensitive, very few people are going to touch this. But I'll add a comment to the field for good measure.

**Comment by agnivade** [comment]
@saturninoabril - This needs to be tested along with https://github.com/mattermost/mattermost-webapp/pull/7921 and https://github.com/mattermost/enterprise/pull/973. 

Please set `"EnableReliableWebSockets": true` in service settings, and test websocket related functionalities for any regression.

Thanks!

**Comment by saturninoabril** [comment]
Thanks @agnivade! I'll wait for https://github.com/mattermost/mattermost-webapp/pull/7921 to be ready for testing. Once ready, please update all the repos and I'll get the build image for testing.

**Comment by agnivade** [comment]
@saturninoabril - You should be good to go now. Thanks!

**Review by saturninoabril — APPROVED** [review]
Tested and passed. Thanks @agnivade!
